### PR TITLE
Networking examples: Call private Cloud Function from onprem

### PR DESCRIPTION
**Pull request by andgandolfi**
Networking examples: Add "Calling a private Cloud Function from On-premises" example

### CODE REVIEW

**Review by juliocc — CHANGES_REQUESTED** [review]
Left a few comments.

File: networking/private-cloud-function-from-onprem/main.tf [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    ttl = 300,
+    records = [var.psc_endpoint]
+  }]
+}
+
+###############################################################################
+#                                  PSCs                                       #
+###############################################################################
+
+resource "google_compute_global_address" "psc-address" {
```
[the remark this answers is not in the document]
Comment (andgandolfi) [reply networking/private-cloud-function-from-onprem/main.tf:270]: Currently the net-address module doesn't support this kind of "google_compute_global_address" resources

File: networking/private-cloud-function-from-onprem/main.tf [hunk 2]
```diff
@@ -251,22 +239,21 @@ module "function-hello" {
 ###############################################################################
 
 module "bucket-functions" {
```
Comment (ludoo) [inline networking/private-cloud-function-from-onprem/main.tf:241]: You can use `bucket-config` in the function module to auto-create bucket, and skip the GCS module.

File: networking/private-cloud-function-from-onprem/main.tf [hunk 3]
```diff
@@ -19,7 +19,7 @@
 ###############################################################################
 
 module "project-onprem" {
-  source          = "github.com/terraform-google-modules/cloud-foundation-fabric//modules/project?ref=v5.0.0"
+  source          = "../../modules/project"
   billing_account = var.billing_account_id
   name            = var.onprem_project_id
```
Comment (ludoo) [inline networking/private-cloud-function-from-onprem/main.tf:24]: I would use a single `projects_ids` variable, with `gcp` and `onprem` attributes.

File: networking/private-cloud-function-from-onprem/main.tf [hunk 4]
```diff
@@ -49,69 +49,60 @@ module "project-gcp" {
...
+  source     = "../../modules/net-vpc"
   project_id = module.project-onprem.project_id
-  name       = "onprem-net"
+  name       = "onprem"
   subnets = [
     {
-      ip_cidr_range      = var.onprem_ip_range
-      name               = "onprem-sub"
+      ip_cidr_range      = var.ip_ranges.onprem
+      name               = "onprem-subnet"
```
Comment (ludoo) [inline networking/private-cloud-function-from-onprem/main.tf:58]: `-subnet` is redundant in a subnet name, but I would use a prefix to make sure th ename does not clash with pre-existing subnets (and also use the prefix for all other resource names).

File: networking/private-cloud-function-from-onprem/main.tf [hunk 5]
```diff
@@ -49,69 +49,60 @@ module "project-gcp" {
...
   ]
 }
 
 module "firewall-onprem" {
-  source               = "github.com/terraform-google-modules/cloud-foundation-fabric//modules/net-vpc-firewall?ref=v5.0.0"
+  source               = "../../modules/net-vpc-firewall"
   project_id           = module.project-onprem.project_id
   network              = module.vpc-onprem.name
   admin_ranges_enabled = true
-  admin_ranges         = ["0.0.0.0/0"]
```
Comment (ludoo) [inline networking/private-cloud-function-from-onprem/main.tf]: You can use the subnet range here, or skip the attribute entirely. You don't need to specify attributes who already have empty defaults.

File: networking/private-cloud-function-from-onprem/main.tf [hunk 6]
```diff
@@ -204,10 +195,10 @@ module "test-vm" {
   network_interfaces = [
     {
       network    = module.vpc-onprem.self_link,
-      subnetwork = module.vpc-onprem.subnet_self_links["${var.region}/onprem-sub"],
+      subnetwork = module.vpc-onprem.subnet_self_links["${var.region}/onprem-subnet"],
       nat        = false,
-      addresses = {
-        internal = [cidrhost(var.onprem_ip_range, 2)]
+      addresses  = {
+        internal = [cidrhost(var.ip_ranges.onprem, 2)]
```
Comment (ludoo) [inline networking/private-cloud-function-from-onprem/main.tf:201]: why do you need to specify the IP?

File: networking/private-cloud-function-from-onprem/main.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (juliocc) [reply networking/private-cloud-function-from-onprem/main.tf:270]: exactly ;)

**Review by andgandolfi — COMMENTED** [review]
Fixed the suggested changes

**Review by ludoo — APPROVED** [review]

**Review by juliocc — APPROVED** [review]

**Comment by juliocc** [comment]
@andgandolfi thanks for the PR

As an enhancement, you can update the net-address module to support PSC (#279) endpoints and use that to reserve the address

**Pull request merged**